src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs

```diff
@@ -394,7 +394,13 @@ private static (DateTime? time, string path) GetLatestInput(IEnumerable<string>
             foreach (string input in inputs)
             {
                 DateTime? time = GetTimestamp(input, timestampCache);
-                if (latest != null && (time == null && !ignoreMissing || time > latest))
+
+                if (time == null && !ignoreMissing)
+                {
+                    return (null, null);
+                }
+
+                if (time > latest)
```

Comment:
I'll say I find this much easier to read.

Reply:
Once I'd convinced myself it was safe, it felt easier to read to me too. Unfortunately other optimisations made things less elegant in other places.